File: en/codes/kotlin/chapter_tree/array_binary_tree.kt
/**
 * File: array_binary_tree.kt
 * Created Time: 2024-01-25
 * Author: curtishd (1023632660@qq.com)
 */

package chapter_tree

import utils.TreeNode
import utils.printTree

/* Binary tree class represented by array */
class ArrayBinaryTree(private val tree: MutableList<Int?>) {
    /* List capacity */
    fun size(): Int {
        return tree.size
    }

    /* Get value of node at index i */
    fun _val(i: Int): Int? {
        // If index out of bounds, return null to represent empty position
        if (i < 0 || i >= size()) return null
        return tree[i]
    }

    /* Get index of left child node of node at index i */
    fun left(i: Int): Int {
        return 2 * i + 1
    }

    /* Get index of right child node of node at index i */
    fun right(i: Int): Int {
        return 2 * i + 2
    }

    /* Get index of parent node of node at index i */
    fun parent(i: Int): Int {
        return (i - 1) / 2
    }

    /* Level-order traversal */
    fun levelOrder(): MutableList<Int?> {
        val res = mutableListOf<Int?>()
        // Traverse array directly
        for (i in 0..<size()) {
            if (_val(i) != null)
                res.add(_val(i))
        }
        return res
    }

    /* Depth-first traversal */
    fun dfs(i: Int, order: String, res: MutableList<Int?>) {
        // If empty position, return
        if (_val(i) == null)
            return
        // Preorder traversal
        if ("pre" == order)
            res.add(_val(i))
        dfs(left(i), order, res)
        // Inorder traversal
        if ("in" == order)
            res.add(_val(i))
        dfs(right(i), order, res)
        // Postorder traversal
        if ("post" == order)
            res.add(_val(i))
    }

    /* Preorder traversal */
    fun preOrder(): MutableList<Int?> {
        val res = mutableListOf<Int?>()
        dfs(0, "pre", res)
        return res
    }

    /* Inorder traversal */
    fun inOrder(): MutableList<Int?> {
        val res = mutableListOf<Int?>()
        dfs(0, "in", res)
        return res
    }

    /* Postorder traversal */
    fun postOrder(): MutableList<Int?> {
        val res = mutableListOf<Int?>()
        dfs(0, "post", res)
        return res
    }
}

/* Driver Code */
fun main() {
    // Initialize binary tree
    // Here we use a function to generate binary tree directly from list
    val arr = mutableListOf(1, 2, 3, 4, null, 6, 7, 8, 9, null, null, 12, null, null, 15)

    val root = TreeNode.listToTree(arr)
    println("\nInitialize binary tree\n")
    println("Array representation of binary tree:")
    println(arr)
    println("Linked list representation of binary tree:")
    printTree(root)

    // Binary tree class represented by array
    val abt = ArrayBinaryTree(arr)

    // Access node
    val i = 1
    val l = abt.left(i)
    val r = abt.right(i)
    val p = abt.parent(i)
    println("Current node index is $i, value is ${abt._val(i)}")
    println("Its left child index is $l, value is ${abt._val(l)}")
    println("Its right child index is $r, value is ${abt._val(r)}")
    println("Its parent node index is $p, value is ${abt._val(p)}")

    // Traverse tree
    var res = abt.levelOrder()
    println("\nLevel-order traversal is: $res")
    res = abt.preOrder()
    println("Pre-order traversal is: $res")
    res = abt.inOrder()
    println("In-order traversal is: $res")
    res = abt.postOrder()
    println("Post-order traversal is: $res")
}

File: en/codes/python/chapter_tree/array_binary_tree.py
"""
File: array_binary_tree.py
Created Time: 2023-07-19
Author: krahets (krahets@163.com)
"""

import sys
from pathlib import Path

sys.path.append(str(Path(__file__).parent.parent))
from modules import TreeNode, list_to_tree, print_tree


class ArrayBinaryTree:
    """Binary tree class represented by array"""

    def __init__(self, arr: list[int | None]):
        """Constructor"""
        self._tree = list(arr)

    def size(self):
        """List capacity"""
        return len(self._tree)

    def val(self, i: int) -> int | None:
        """Get value of node at index i"""
        # If index is out of bounds, return None, representing empty position
        if i < 0 or i >= self.size():
            return None
        return self._tree[i]

    def left(self, i: int) -> int | None:
        """Get index of left child node of node at index i"""
        return 2 * i + 1

    def right(self, i: int) -> int | None:
        """Get index of right child node of node at index i"""
        return 2 * i + 2

    def parent(self, i: int) -> int | None:
        """Get index of parent node of node at index i"""
        return (i - 1) // 2

    def level_order(self) -> list[int]:
        """Level-order traversal"""
        self.res = []
        # Traverse array directly
        for i in range(self.size()):
            if self.val(i) is not None:
                self.res.append(self.val(i))
        return self.res

    def dfs(self, i: int, order: str):
        """Depth-first traversal"""
        if self.val(i) is None:
            return
        # Preorder traversal
        if order == "pre":
            self.res.append(self.val(i))
        self.dfs(self.left(i), order)
        # Inorder traversal
        if order == "in":
            self.res.append(self.val(i))
        self.dfs(self.right(i), order)
        # Postorder traversal
        if order == "post":
            self.res.append(self.val(i))

    def pre_order(self) -> list[int]:
        """Preorder traversal"""
        self.res = []
        self.dfs(0, order="pre")
        return self.res

    def in_order(self) -> list[int]:
        """Inorder traversal"""
        self.res = []
        self.dfs(0, order="in")
        return self.res

    def post_order(self) -> list[int]:
        """Postorder traversal"""
        self.res = []
        self.dfs(0, order="post")
        return self.res


"""Driver Code"""
if __name__ == "__main__":
    # Initialize binary tree
    # Here we use a function to generate a binary tree directly from an array
    arr = [1, 2, 3, 4, None, 6, 7, 8, 9, None, None, 12, None, None, 15]
    root = list_to_tree(arr)
    print("\nInitialize binary tree\n")
    print("Array representation of binary tree:")
    print(arr)
    print("Linked list representation of binary tree:")
    print_tree(root)

    # Binary tree class represented by array
    abt = ArrayBinaryTree(arr)

    # Access nodes
    i = 1
    l, r, p = abt.left(i), abt.right(i), abt.parent(i)
    print(f"\nCurrent node's index is {i}, value is {abt.val(i)}")
    print(f"Its left child node's index is {l}, value is {abt.val(l)}")
    print(f"Its right child node's index is {r}, value is {abt.val(r)}")
    print(f"Its parent node's index is {p}, value is {abt.val(p)}")

    # Traverse tree
    res = abt.level_order()
    print("\nLevel-order traversal is:", res)
    res = abt.pre_order()
    print("Preorder traversal is:", res)
    res = abt.in_order()
    print("Inorder traversal is:", res)
    res = abt.post_order()
    print("Postorder traversal is:", res)
